roachpb: introduce QueryIntent request

nvb opened the pull request:
This change introduces a new request method called QueryIntent. The request
type draws parallels to the QueryTxn method, but it visits an intent instead
of a transaction record and returns whether the intent exists or not. The
request type also includes an "if missing" behavior, which allows clients
to optionally ensure that the request prevents a missing intent from ever
being written or return an error if the intent is found to be missing.

This request type was proposed/discussed in both #24194 and #16026. It is a
prerequisite for either proposal.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/26335)
<!-- Reviewable:end -->

bdarnell commented:
Review status: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---

*[pkg/roachpb/api.go, line 1060 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26335#-LEAuPYQ6hlgxJgDV4Wy:-LEAuPYQ6hlgxJgDV4Wz:b6pa53j) ([raw file](https://github.com/cockroachdb/cockroach/blob/c3b6e33acf0394e62f22c2f03f3dc0a3d9de3d3c/pkg/roachpb/api.go#L1060)):*
> ```Go
> func (*PushTxnRequest) flags() int               { return isWrite | isAlone }
> func (*QueryTxnRequest) flags() int              { return isRead | isAlone }
> func (*QueryIntentRequest) flags() int           { return isRead | updatesReadTSCache }
> ```

Comment that this only updates the read TS cache sometimes (and a reference to the code below)

---

*[pkg/roachpb/api.proto, line 797 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26335#-LEAs0R17Bc2tq1vnT5Y:-LEAs0R27R0qyKEJIoPq:b-8jstwr) ([raw file](https://github.com/cockroachdb/cockroach/blob/c3b6e33acf0394e62f22c2f03f3dc0a3d9de3d3c/pkg/roachpb/api.proto#L797)):*
> ```protobuf
>     // response.
>     DO_NOTHING = 0;
>     // Throw an IntentMissingError.
> ```

s/throw/return/

---

*[pkg/roachpb/api.proto, line 810 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26335#-LEAs8o83qwN2PI1U8n1:-LEAs8o83qwN2PI1U8n2:b4c9lyr) ([raw file](https://github.com/cockroachdb/cockroach/blob/c3b6e33acf0394e62f22c2f03f3dc0a3d9de3d3c/pkg/roachpb/api.proto#L810)):*
> ```protobuf
>   // behavior could be a function of (BatchRequest.Txn.ID == ExpectedTxn.ID).
>   // This might be convenient, but I think it's probably better in the long run
>   // to keep this behavior explicit and orthogonal to the state of the caller.
> ```

:thumbsup: 

---

*[pkg/roachpb/api.proto, line 816 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26335#-LEAsPzH-3xr-7fbmmYi:-LEAsPzH-3xr-7fbmmYj:bawn1f2) ([raw file](https://github.com/cockroachdb/cockroach/blob/c3b6e33acf0394e62f22c2f03f3dc0a3d9de3d3c/pkg/roachpb/api.proto#L816)):*
> ```protobuf
>   // It also might make more sense to make if_missing_throw_error and
>   // if_missing_prevent independent fields so that they become completely
>   // orthogonal options that can be used together or independently.
> ```

I'm not sure we would ever want both - if an error is returned, the rest of the batch is failed, so we wouldn't reliably prevent all the intents if any of them encountered an error. 

---

*[pkg/roachpb/api.proto, line 825 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26335#-LEAsgnl22nPdSl1zAWR:-LEAsgnl22nPdSl1zAWS:b3wt73w) ([raw file](https://github.com/cockroachdb/cockroach/blob/c3b6e33acf0394e62f22c2f03f3dc0a3d9de3d3c/pkg/roachpb/api.proto#L825)):*
> ```protobuf
>   // Whether an intent matching the expected transaction was found at the key.
>   //
>   // TODO DURING REIVEW: we could return the entire TxnMeta here instead. This
> ```

I think it's fine to not return a TxnMeta, since the caller is required to pass in a TxnMeta. 

Does a `bool` capture all possibilities? Do we need to distinguish between "not found, but could still be written" from "not found, and could no longer be written"?

---

*[pkg/roachpb/errors.proto, line 384 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26335#-LEAtblTAg02lJeRIYgm:-LEAtblTAg02lJeRIYgn:b-pxkekj) ([raw file](https://github.com/cockroachdb/cockroach/blob/c3b6e33acf0394e62f22c2f03f3dc0a3d9de3d3c/pkg/roachpb/errors.proto#L384)):*
> ```protobuf
>   option (gogoproto.equal) = true;
> 
>   optional storage.engine.enginepb.TxnMeta txn = 1 [(gogoproto.nullable) = false];
> ```

Document this. Why do we echo the caller's TxnMeta back to them? If we need a TxnMeta here, maybe we should include one in the QueryIntentResponse.

---

*[pkg/storage/replica.go, line 2089 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26335#-LEAwElY4JgRUaunlImu:-LEAwElY4JgRUaunlImv:b94auoo) ([raw file](https://github.com/cockroachdb/cockroach/blob/c3b6e33acf0394e62f22c2f03f3dc0a3d9de3d3c/pkg/storage/replica.go#L2089)):*
> ```Go
> 						// TODO DURING REVIEW: do we want to update the write timestamp
> 						// cache instead so that we ensure that the WriteToOld flag
> 						// is set on the eventual intent write?
> ```

No, I don't think setting WriteTooOld is required here. WriteTooOld forces transaction restarts because there is known to be a conflicting write; to prevent the writing of an intent the read timestamp cache is sufficient and a bit cheaper.

---

*[pkg/storage/batcheval/cmd_query_intent.go, line 59 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26335#-LEAz6aaD5GiwQ2N0o5L:-LEAz6aaD5GiwQ2N0o5M:b-ct1xe8) ([raw file](https://github.com/cockroachdb/cockroach/blob/c3b6e33acf0394e62f22c2f03f3dc0a3d9de3d3c/pkg/storage/batcheval/cmd_query_intent.go#L59)):*
> ```Go
> 		i := intents[0]
> 		reply.FoundIntent = (i.Txn.ID == args.Txn.ID) &&
> 			(i.Txn.Epoch >= args.Txn.Epoch) &&
> ```

Is the `>=` required here? When would the epoch and sequence not be an exact match?

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26335)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[pkg/roachpb/api.go, line 1060 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26335#-LEAuPYQ6hlgxJgDV4Wy:-LER55zFAWd-ZDKKOW4n:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c3b6e33acf0394e62f22c2f03f3dc0a3d9de3d3c/pkg/roachpb/api.go#L1060)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Comment that this only updates the read TS cache sometimes (and a reference to the code below)
</blockquote></details>

Done.

---

*[pkg/roachpb/api.proto, line 797 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26335#-LEAs0R17Bc2tq1vnT5Y:-LEQnkRO5PtEzXGlEOm-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c3b6e33acf0394e62f22c2f03f3dc0a3d9de3d3c/pkg/roachpb/api.proto#L797)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

s/throw/return/
</blockquote></details>

Done.

---

*[pkg/roachpb/api.proto, line 816 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26335#-LEAsPzH-3xr-7fbmmYi:-LEQndOdCETKFRYwlx_y:b-q7loh1) ([raw file](https://github.com/cockroachdb/cockroach/blob/c3b6e33acf0394e62f22c2f03f3dc0a3d9de3d3c/pkg/roachpb/api.proto#L816)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

I'm not sure we would ever want both - if an error is returned, the rest of the batch is failed, so we wouldn't reliably prevent all the intents if any of them encountered an error. 
</blockquote></details>

Good point. I'll keep this as is.

---

*[pkg/roachpb/api.proto, line 825 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26335#-LEAsgnl22nPdSl1zAWR:-LEQolq77FUMwk2GxGcg:b-fo2jk6) ([raw file](https://github.com/cockroachdb/cockroach/blob/c3b6e33acf0394e62f22c2f03f3dc0a3d9de3d3c/pkg/roachpb/api.proto#L825)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

I think it's fine to not return a TxnMeta, since the caller is required to pass in a TxnMeta. 

Does a `bool` capture all possibilities? Do we need to distinguish between "not found, but could still be written" from "not found, and could no longer be written"?
</blockquote></details>

I don't think we need to make this distinction, at least for any of our proposed uses of this request. If a client needs to prevent a missing intent then it will use the `PREVENT` behavior, in which case this bool indicates the latter. If a client doesn't need to prevent the intent, then it currently doesn't cares about the difference.

---

*[pkg/roachpb/errors.proto, line 384 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26335#-LEAtblTAg02lJeRIYgm:-LEQq3w2-2t1fvrQ_4pG:b8r6att) ([raw file](https://github.com/cockroachdb/cockroach/blob/c3b6e33acf0394e62f22c2f03f3dc0a3d9de3d3c/pkg/roachpb/errors.proto#L384)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Document this. Why do we echo the caller's TxnMeta back to them? If we need a TxnMeta here, maybe we should include one in the QueryIntentResponse.
</blockquote></details>

Done. And added the incorrect intent that was found, if any, to give the error message more color.

---

*[pkg/storage/replica.go, line 2089 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26335#-LEAwElY4JgRUaunlImu:-LEQtc933WjjxVIhPI20:b-fucsey) ([raw file](https://github.com/cockroachdb/cockroach/blob/c3b6e33acf0394e62f22c2f03f3dc0a3d9de3d3c/pkg/storage/replica.go#L2089)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

No, I don't think setting WriteTooOld is required here. WriteTooOld forces transaction restarts because there is known to be a conflicting write; to prevent the writing of an intent the read timestamp cache is sufficient and a bit cheaper.
</blockquote></details>

This depends on what we mean by "prevent the writing of an intent". I'm realizing (see `TestQueryIntentRequest`) that the read timestamp cache doesn't prevent the intent entirely, it just forces it to bump its timestamp. That's sufficient in the case of parallel commits because `DistSender` will check whether the transaction had its timestamp pushed by any promised write before committing the transaction.

Removed this comment.

---

*[pkg/storage/batcheval/cmd_query_intent.go, line 59 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26335#-LEAz6aaD5GiwQ2N0o5L:-LEQyiEQ0a37UAv7bZ_Y:b-799j1s) ([raw file](https://github.com/cockroachdb/cockroach/blob/c3b6e33acf0394e62f22c2f03f3dc0a3d9de3d3c/pkg/storage/batcheval/cmd_query_intent.go#L59)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Is the `>=` required here? When would the epoch and sequence not be an exact match?
</blockquote></details>

The epoch should have been `==`. This was also missing the very important `Timestamp` comparison, which is necessary because an intent that has a larger timestamp than expected must have been pushed and should not count as a matching intent.

The sequence number is the only interesting case where we want more than exact equality. See the [parallel commits RFC](https://github.com/tschottdorf/cockroach/blob/a2a46fce6d36f376b20868ed8283a14766436275/docs/RFCS/20180324_parallel_commit.md#preventintentrequest) or the command I added to `QueryIntentRequest.Txn` for a discussion on why.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26335)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
Review status: 0 of 13 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[pkg/storage/batcheval/cmd_query_intent.go, line 59 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26335#-LEAz6aaD5GiwQ2N0o5L:-LEWi_VA9CDxy_9Dt_eF:b-8q5sca) ([raw file](https://github.com/cockroachdb/cockroach/blob/c3b6e33acf0394e62f22c2f03f3dc0a3d9de3d3c/pkg/storage/batcheval/cmd_query_intent.go#L59)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

The epoch should have been `==`. This was also missing the very important `Timestamp` comparison, which is necessary because an intent that has a larger timestamp than expected must have been pushed and should not count as a matching intent.

The sequence number is the only interesting case where we want more than exact equality. See the [parallel commits RFC](https://github.com/tschottdorf/cockroach/blob/a2a46fce6d36f376b20868ed8283a14766436275/docs/RFCS/20180324_parallel_commit.md#preventintentrequest) or the command I added to `QueryIntentRequest.Txn` for a discussion on why.
</blockquote></details>

I think the Timestamp comparison is off here and in the parallel commits RFC. I think it actually should be checking that the intent has a timestamp that is "less than or equal to" the txn's provisional commit timestamp. The reason for this is that a txn's intents don't need to be at its commit timestamp to allow it to commit, they just need to be at a timestamp equal to or less than its commit timestamp.

This comes up with write pipelining because a batch may have its timestamp bumped after it laid down intents but from the point of view of the txn coordinator it's unclear which of its intents were laid down at which timestamp. If the txn refreshes successfully then its intents, even at older timestamps, are still valid and a QueryIntent request should still see them.

Likewise, for parallel commits, a pusher trying to prevent a transaction from committing needs to prove that the transaction has failed to lay down an intent that it promises to write before committing. If the pusher finds an intent with the same sequence number but a smaller timestamp then bumping the timestamp cache won't do anything from the point of view of the committing transaction. In its eyes, it has laid down a "committable" intent.

cc. @tschottdorf to double check this reasoning.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26335)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Nathan, that checks out. Thanks for noticing.

On Sat, Jun 9, 2018, 1:38 AM Nathan VanBenschoten <notifications@github.com>
wrote:

> Review status: 0 of 13 files reviewed at latest revision, 8 unresolved
> discussions, some commit checks failed.
> ------------------------------
>
> *pkg/storage/batcheval/cmd_query_intent.go, line 59 at r1
> <https://reviewable.io/reviews/cockroachdb/cockroach/26335#-LEAz6aaD5GiwQ2N0o5L:-LEWi_VA9CDxy_9Dt_eF:b-8q5sca>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/c3b6e33acf0394e62f22c2f03f3dc0a3d9de3d3c/pkg/storage/batcheval/cmd_query_intent.go#L59>):*
> *Previously, nvanbenschoten (Nathan VanBenschoten) wrote…*
>
> The epoch should have been ==. This was also missing the very important
> Timestamp comparison, which is necessary because an intent that has a
> larger timestamp than expected must have been pushed and should not count
> as a matching intent.
>
> The sequence number is the only interesting case where we want more than
> exact equality. See the parallel commits RFC
> <https://github.com/tschottdorf/cockroach/blob/a2a46fce6d36f376b20868ed8283a14766436275/docs/RFCS/20180324_parallel_commit.md#preventintentrequest>
> or the command I added to QueryIntentRequest.Txn for a discussion on why.
>
> I think the Timestamp comparison is off here and in the parallel commits
> RFC. I think it actually should be checking that the intent has a timestamp
> that is "less than or equal to" the txn's provisional commit timestamp. The
> reason for this is that a txn's intents don't need to be at its commit
> timestamp to allow it to commit, they just need to be at a timestamp equal
> to or less than its commit timestamp.
>
> This comes up with write pipelining because a batch may have its timestamp
> bumped after it laid down intents but from the point of view of the txn
> coordinator it's unclear which of its intents were laid down at which
> timestamp. If the txn refreshes successfully then its intents, even at
> older timestamps, are still valid and a QueryIntent request should still
> see them.
>
> Likewise, for parallel commits, a pusher trying to prevent a transaction
> from committing needs to prove that the transaction has failed to lay down
> an intent that it promises to write before committing. If the pusher finds
> an intent with the same sequence number but a smaller timestamp then
> bumping the timestamp cache won't do anything from the point of view of the
> committing transaction. In its eyes, it has laid down a "committable"
> intent.
>
> cc. @tschottdorf <https://github.com/tschottdorf> to double check this
> reasoning.
> ------------------------------
>
> *Comments from Reviewable
> <https://reviewable.io/reviews/cockroachdb/cockroach/26335>*
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/26335#issuecomment-395919700>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AE135N3jUbt_jrSY7elv1JLLIArR8wTZks5t6wsKgaJpZM4UXV6g>
> .
>
-- 

-- Tobias

bdarnell commented:
Reviewed 13 of 13 files at r2.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[pkg/roachpb/api.proto, line 797 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26335#-LEAs0R17Bc2tq1vnT5Y:-LEausj8BbeRUHHjYYDs:b-ojzxto) ([raw file](https://github.com/cockroachdb/cockroach/blob/c3b6e33acf0394e62f22c2f03f3dc0a3d9de3d3c/pkg/roachpb/api.proto#L797)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Change the constant too. 

---

*[pkg/roachpb/api.proto, line 749 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/26335#-LEawUo32IhWOALexu7j:-LEawUo32IhWOALexu7k:b-rfl8eb) ([raw file](https://github.com/cockroachdb/cockroach/blob/42df0a83791c513be5916d1068204aaf5d83247d/pkg/roachpb/api.proto#L749)):*
> ```protobuf
>   // the specified key, the intent is only considered a match if it has the same
>   // ID, the same epoch, and the same provisional commit timestamp. The
>   // timestamp check enforces that if a transaction was pushed before laying
> ```

Update this comment with the new rules for timestamp matches.

---

*[pkg/roachpb/errors.proto, line 384 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26335#-LEAtblTAg02lJeRIYgm:-LEawnN-0R426nlcMX1b:b3k1523) ([raw file](https://github.com/cockroachdb/cockroach/blob/c3b6e33acf0394e62f22c2f03f3dc0a3d9de3d3c/pkg/roachpb/errors.proto#L384)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Done. And added the incorrect intent that was found, if any, to give the error message more color.
</blockquote></details>

Why do we need to return this at all since it was a part of the request?

---

*[pkg/storage/replica.go, line 2089 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26335#-LEAwElY4JgRUaunlImu:-LEawzp43EX6trbQKkHL:b81a6s1) ([raw file](https://github.com/cockroachdb/cockroach/blob/c3b6e33acf0394e62f22c2f03f3dc0a3d9de3d3c/pkg/storage/replica.go#L2089)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

This depends on what we mean by "prevent the writing of an intent". I'm realizing (see `TestQueryIntentRequest`) that the read timestamp cache doesn't prevent the intent entirely, it just forces it to bump its timestamp. That's sufficient in the case of parallel commits because `DistSender` will check whether the transaction had its timestamp pushed by any promised write before committing the transaction.

Removed this comment.
</blockquote></details>

Yes, I think we just mean "prevent writing an intent at this timestamp", so the read timestamp cache is sufficient. 

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26335)*
<!-- Sent from Reviewable.io -->

nvb committed:
roachpb: introduce QueryIntent request

This change introduces a new request method called QueryIntent. The request
type draws parallels to the QueryTxn method, but it visits an intent instead
of a transaction record and returns whether the intent exists or not. The
request type also includes an "if missing" behavior, which allows clients
to optionally ensure that the request prevents a missing intent from ever
being written or return an error if the intent is found to be missing.

This request type was proposed/discussed in both #24194 and #16026. It is a
prerequisite for either proposal.

Release note: None

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 8 of 8 files at r3.
Review status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

---



*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26335#-:-LFrcl8c3TOcCmLjXnx5:bnfp4nl)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
TFTR.

bors r=bdarnell
